### ISSUE TITLE
Priority of overlapping groups: latex brackets

### ISSUE THREAD

**Issue by kiryph** [bootleq/vim-cycle#5]
Consider following groups of latex-brackets (see details http://ctan.space-pro.be/tex-archive/obsolete/info/math/voss/mathmode/Mathmode.pdf#page=24)

``` viml
let g:cycle_default_groups_for_tex = [
        \   [['\big(:\big)', '\Big(:\Big)', '\bigg(:\bigg)', '\Bigg(:\Bigg)'], 'sub_pairs', {'hard_case':1}, {'match_case':1}],
        \   [['\big[:\big]', '\Big[:\Big]', '\bigg[:\bigg]', '\Bigg[:\Bigg]'], 'sub_pairs', {'hard_case':1}, {'match_case':1}],
        \   [['\big\{:\big\}', '\Big\{:\Big\}', '\bigg\{:\bigg\}', '\Bigg\{:\Bigg\}'], 'sub_pairs', {'hard_case':1}, {'match_case':1}],
        \   [['\big|:\big|', '\Big|:\Big|', '\bigg|:\bigg|', '\Bigg|:\Bigg|'], 'sub_pairs', {'hard_case':1}, {'match_case':1}],
        \   [['\big\Vert:\big\Vert', '\Big\Vert:\Big\Vert', '\bigg\Vert:\bigg\Vert', '\Bigg\Vert:\Bigg\Vert'], 'sub_pairs', {'hard_case':1}, {'match_case':1}],
        \   [['\big\lfloor:\big\rfloor', '\Big\lfloor:\Big\rfloor', '\bigg\lfloor:\bigg\rfloor', '\Bigg\lfloor:\Bigg\rfloor'], 'sub_pairs', {'hard_case':1}, {'match_case':1}],
        \   [['\big\lceil:\big\rceil', '\Big\lceil:\Big\rceil', '\bigg\lceil:\bigg\rceil', '\Bigg\lceil:\Bigg\rceil'], 'sub_pairs', {'hard_case':1}, {'match_case':1}],
        \   [['\big\langle:\big\rangle', '\Big\langle:\Big\rangle', '\bigg\langle:\bigg\rangle', '\Bigg\langle:\Bigg\rangle'], 'sub_pairs', {'hard_case':1}, {'match_case':1}],
        \   [['\big', '\Big', '\bigg', '\Bigg'], {'hard_case':1}, {'match_case':1}],
        \   [['(:)', '[:]', '\{:\}', '\lfloor :\rfloor', '\lceil :\rceil', '\langle :\rangle'], 'sub_pairs'],
        \ ]
```

Consider following example tex file:

``` tex
\big\langle A+B \big\rangle
```

I'd like to change type of brackets when the cursor is on the bracket (e.g. `\rangle`) and I'd like to change the size of the bracket when cursor is on on the size specifier (e.g. `\big`).

And I would like to have a fallback behaviour for

``` tex
\big\langle A \big| B \big\rangle
```

Is this possible? I admit that this stretches the capabilities of this plugin. I don't know if https://github.com/AndrewRadev/switch.vim could do this.

**Comment by kiryph** [comment]
I have noticed that following two groups are already not working:

``` viml
let g:cycle_default_groups_for_tex = [
  \   [['\big(:\big)', '\Big(:\Big)', '\bigg(:\bigg)', '\Bigg(:\Bigg)'], 'sub_pairs', {'hard_case':1}, {'match_case':1}],
  \   [['\big[:\big]', '\Big[:\Big]', '\bigg[:\bigg]', '\Bigg[:\Bigg]'], 'sub_pairs', {'hard_case':1}, {'match_case':1}],
  \ ]
```

``` viml
:echo b:cycle_groups
[{'options': {'sub_pair': 1, 'end_with': ['\big)', '\Big)', '\bigg)', '\Bigg)'], 'match_case': 1, 'hard_case': 1}, 'items': ['\big(', '\Big(', '\bigg(', '\Bigg(']},
 {'options': {'sub_pair': 1, 'begin_with': ['\big(', '\Big(', '\bigg(', '\Bigg('], 'match_case': 1, 'hard_case': 1, 'restrict_cursor': 1}, 'items': ['\big)', '\Big)', '\bigg)', '\Bigg)']},
 {'options': {'sub_pair': 1, 'end_with': ['\big]', '\Big]', '\bigg]', '\Bigg]'], 'match_case': 1, 'hard_case': 1}, 'items': ['\big[', '\Big[', '\bigg[', '\Bigg[']},
 {'options': {'sub_pair': 1, 'begin_with': ['\big[', '\Big[', '\bigg[', '\Bigg['], 'match_case': 1, 'hard_case': 1}, 'items': ['\big]', '\Big]', '\bigg]', '\Bigg]']}]  
```

It is not possible to increase the bracket size automatically for opening and closing bracket

```
\big( a+b \big)
```

After pressing once `ctrl-a` it should become

```
\Big( a+b \Big)
```

but instead becomes

```
\Big( a+b \big)
```

**Comment by bootleq** [comment]
Thanks for detailed report, just found a bug during change ending `\big)` to `\Big)`, which should be fixed in v0.2.1, and I added some example setting similar to your use case:

```
    \big\langle A+B \big\rangle
```

Please see if [group_example#L18-L25](https://github.com/bootleq/vim-cycle/blob/f4e3f31a4c57b4997bd6a55d83f7b4340ee56e4f/test/group_operation.vim#L18-L25) is ok to change the size and bracket.
While there is a known limitation begin/end items can't be the same, so `\big|:\big|` pair can't work for the time being.

For the fallback example you provided:

```
    \big\langle A \big| B \big\rangle
```

I'm not sure I understand it, is it cycle `+` to `\big`?
Anyway it seems not closely related to group priority issue, so I guess we can close this issue if no rejection.

And about switch.vim, Yes I believe it can do very clever setting to archive more advanced usage, it looks great but I didn't try use it yet.

**Comment by kiryph** [comment]
Thank you very much for your update.

### v0.2.1 fixes posted issue

In the equation`\big\langle A+B \big\rangle` the _bracket size_ can now be changed when the cursor is positioned either on the left or the right `\big`. And you can change the _bracket type_ when the cursor is either on `\langle` or `\rangle`.

### `\lVert`/`\rVert` and `\lvert`/`\rvert`

Your example definition contains `\lVert` and `\rVert` which is a nice way to integrate `\Vert` into this workflow. However, they do require the package `amsmath`. This reminded me that there is also `\lvert` and `\rvert` which can be used for `|`. To improve compactness of the tex-code, one can enable concealment with `:se cole=2`. Unfortunately, `lvert` and `rVert` do not work out of the box:
Add following to `~/.vim/after/syntax/tex.vim` to define the concealment character

``` viml
" Amsmath [lr][vV]ert
let s:texMathDelimList=[
     \ ['\\lvert'     , '|'] ,
     \ ['\\rvert'     , '|'] ,
     \ ['\\lVert'     , '‖'] ,
     \ ['\\rVert'     , '‖'] ,
     \ ]
for texmath in s:texMathDelimList
    exe "syn match texMathDelim '\\\\[bB]igg\\=[lr]\\=".texmath[0]."'   contained conceal cchar=".texmath[1]
endfor
```

My group definition for tex looks like this.

``` viml
" requires amsmath for `Vert` and `vert`
let g:cycle_default_groups_for_tex = [
        \   [['tiny', 'scriptsize', 'footnotesize', 'small', 'normalsize', 'large', 'Large', 'LARGE', 'huge', 'Huge'], 'hard_case', 'match_case'],
        \   [['displaystyle', 'scriptstyle', 'scriptscriptstyle', 'textstyle']],
        \   [['part', 'chapter', 'section', 'subsection', 'subsubsection', 'paragraph', 'subparagraph']],
        \   [['article', 'report', 'book', 'letter', 'slides']],
        \   [['a4paper', 'a5paper', 'b5paper', 'executivepaper', 'legalpaper', 'letterpaper']],
        \   [['scrbook', 'scrreprt', 'scrartcl', 'scrlttr2']],
        \   [['oneside', 'twoside']],
        \   [['onecolumn', 'twocolumn']],
        \   [['draft', 'final']],
        \   [['(:)', '[:]', '\{:\}', '\lvert :\rvert', '\lVert :\rVert', '\lfloor :\rfloor', '\lceil :\rceil', '\langle :\rangle'], 'sub_pairs', 'hard_case', 'match_case'],
        \   [['\big(:\big)', '\Big(:\Big)', '\bigg(:\bigg)', '\Bigg(:\Bigg)'], 'sub_pairs', 'hard_case', 'match_case'],
        \   [['\big[:\big]', '\Big[:\Big]', '\bigg[:\bigg]', '\Bigg[:\Bigg]'], 'sub_pairs', 'hard_case', 'match_case'],
        \   [['\big\{:\big\}', '\Big\{:\Big\}', '\bigg\{:\bigg\}', '\Bigg\{:\Bigg\}'], 'sub_pairs', 'hard_case', 'match_case'],
        \   [['\big\l:\big\r', '\Big\l:\Big\r', '\bigg\l:\bigg\r', '\Bigg\l:\Bigg\r'], 'sub_pairs', 'hard_case', 'match_case'],
        \   [['\big', '\Big', '\bigg', '\Bigg'], 'hard_case', 'match_case'],
        \ ]
```

 **Note:** It is important that the bracket pairs come before the bracket sizes. Otherwise, when the cursor is over `\l` or `\r` the bracket size is cycled instead of the bracket type which I wouldn't expect.

### Fallback behavior

I wasn't very clear what I meant with fallback behavior. I was referring to adjusting bracket sizes in following situations:
1.  a third \big occurs in `\big( A \bigm| B \big)` (auto-size version `\left( A \middle| B \right)`)
2. or an asymmetric version `\big. A \big)`

Placing the group `[['\big', '\Big', '\bigg', '\Bigg'], 'hard_case', 'match_case']` at the very end of the group definitions makes it possible to cycle single bracket size specifier.

Maybe in the future I would consider finding a solution to handle even them automatically. But right now I am very happy to see this working with your plugin. Thanks for your time.
